File: agent-workflow/ui/src/protocol.ts
// The event format the UI understands (protocol version 1).
// Python mirror: agent-workflow/server/protocol.py. Keep the two in sync.
//
// A run is a sequence of events:  meta → workflow → (step | snapshot)* → end
// Any data source can drive the UI as long as it produces these (see sources.ts).

export const PROTOCOL_VERSION = 1;

/** Built-in kinds have their own badge color. Any other string works and gets a neutral badge. */
export type StepKind = "trigger" | "code" | "ai" | "human" | (string & {});
export type StepStatus = "idle" | "running" | "done" | "flagged";

export type StepDefinition = { id: string; kind: StepKind; title: string; role?: string };
export type EdgeDefinition = { source: string; target: string };
export type WorkflowDefinition = { steps: StepDefinition[]; edges: EdgeDefinition[] };

export type StepResult = {
  summary: string;
  input: string[];
  output: string[];
  reasoning: string | null;
};

export type UnitSnapshot = { code: string; name: string; occupied: number; capacity: number; waiting: number };
export type Snapshot = { label: string; clock: string; units: UnitSnapshot[] };

export type MetaEvent = { type: "meta"; protocol: number; seed?: number; [key: string]: unknown };

export type SwarmEvent =
  /** First event. Anything extra (seed, casualties…) is kept in `swarm.meta`. */
  | MetaEvent
  /** The graph to draw. Optional: without it the UI keeps the graph it already has. */
  | ({ type: "workflow" } & WorkflowDefinition)
  /** A step started (`running`) or finished (`done`, or `flagged` when a person must check it). */
  | ({ type: "step"; id: string; status: Exclude<StepStatus, "idle"> } & Partial<StepResult>)
  /** Bed counts per unit at a point in time. */
  | ({ type: "snapshot" } & Snapshot)
  /** Something went wrong; the run stops. */
  | { type: "error"; message: string }
  /** Last event. */
  | { type: "end" };

File: agent-workflow/server/protocol.py
"""The event format the UI understands (protocol version 1).

Any data source — this simulation, a Gemini-backed pipeline, n8n, a partner's
service — can drive the UI by emitting these events in this order:

    meta → workflow → (step running/done, snapshot)* → end

The TypeScript mirror of these shapes is agent-workflow/ui/src/protocol.ts.
"""

PROTOCOL_VERSION = 1

STEP_KINDS = ("trigger", "code", "ai", "human")
STEP_STATUSES = ("running", "done", "flagged")


def meta(**info):
    """First event of a run. Put anything useful to display here (e.g. seed)."""
    return {"type": "meta", "protocol": PROTOCOL_VERSION, **info}


def workflow(steps, edges):
    """The graph the UI draws. steps: [{id, kind, title, role}], edges: [(source, target)]."""
    return {
        "type": "workflow",
        "steps": [dict(s) for s in steps],
        "edges": [{"source": a, "target": b} for a, b in edges],
    }


def running(*step_ids):
    """One or more steps started. Several at once means they run in parallel."""
    return [{"type": "step", "id": i, "status": "running"} for i in step_ids]


def step_done(step_id, summary, input, output, reasoning=None, flagged=False):
    """A step finished. `flagged` means a person needs to look at it."""
    return {
        "type": "step",
        "id": step_id,
        "status": "flagged" if flagged else "done",
        "summary": summary,
        "input": list(input),
        "output": list(output),
        "reasoning": reasoning,
    }


def snapshot(label, clock, units):
    """Bed counts. units: [{code, name, occupied, capacity, waiting}]."""
    return {"type": "snapshot", "label": label, "clock": clock, "units": units}


def error(message):
    return {"type": "error", "message": message}


def end():
    return {"type": "end"}


def pause(seconds):
    """Not sent to the UI. Tells the server to wait so a person can follow along."""
    return {"type": "pause", "seconds": seconds}
